### PR TITLE
Navier-Stokes operator

### PR DESCRIPTION
**Pull request by MTCam**
Adds the Navier-Stokes operator used in production, including @thomasgibson 's overintegration additions.

Change locations summary:

Docs: `*.rst`  Updated just to reflect the module is added, NS docs are already in main
NS operator: `navierstokes.py`, with some supporting ops in `operator.py` and `viscous.py`, includes @thomasgibson overintegration with tests and examples (`test_navierstokes.py`, `test_operators.py`), (`poiseuillle-mpi.py`, `hotplate-mpi.py`, `nsmix-mpi.py`)  
Viscous boundary conditions: `boundary.py`  (**)
Tests updated for new interfaces: `test_*.py`
Supporting updates:
  - Adds state creation utility to unify/share code among fluid (and multiphysics) operators: `gas_mode.py`
  - Euler operator updated to use state creation util: `euler.py`
  - Centralize discretization creation (saves a lot of time updating examples to use Discr collection): `discretization.py`
  - Unification of EOS interface mix v. single gas: `eos.py`
  - Address deprecation/data structure warning with CV array container: `fluid.py`
  - Adds symbolic infrastructure for testing fluid components: `symbolic_fluid.py`
  - Lazy transport bug fix for transport model return types: `transport.py`

 
(**) Note that this change set extends the current boundary treatments and mechanisms to handle viscous boundary conditions as needed by the Navier-Stokes operator.   It is not the end of the boundary treatment story, as we plan on a significant refactor and upcoming wall-model coupling.

**Questions for the review**:
- [x] Is the scope and purpose of the PR clear?
  - [x] The PR should have a description.
  - [x] The PR should have a guide if needed (e.g., an ordering).
- [ ] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

### CODE REVIEW

File: mirgecom/flux.py [hunk 1]
```diff
@@ -35,6 +44,40 @@
...
+    f_minus = physical_flux_func(state_pair.int)@normal
+    f_plus = physical_flux_func(state_pair.ext)@normal
+    q_minus = state_pair.int.cv
+    q_plus = state_pair.ext.cv
+    f_star = (s_plus*f_minus - s_minus*f_plus
+              + s_plus*s_minus*(q_plus - q_minus))/(s_plus - s_minus)
+
+    # choose the correct f contribution based on the wave speeds
+    f_check_minus = actx.np.greater_equal(s_minus, zeros)*(0*f_minus + 1.0)
+    f_check_plus = actx.np.less_equal(s_plus, zeros)*(0*f_minus + 1.0)
```
Comment (thomasgibson) [inline mirgecom/flux.py:72]: I understand why it was written like this (broadcasting issues with pyopencl, lazy (pytato), and numpy-like data structures), but I think we can probably simplify this once we pin down a good array context for testing (I'm thinking numpy is the way to go, and some work has started on it: https://github.com/inducer/arraycontext/pull/93).

File: mirgecom/inviscid.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 .. autofunction:: inviscid_flux
 .. autofunction:: inviscid_facial_flux
+.. autofunction:: inviscid_flux_rusanov
```
Comment (thomasgibson) [inline mirgecom/inviscid.py:8]:
```suggestion
.. autofunction:: inviscid_flux_rusanov
.. autofunction:: inviscid_flux_hll
```

File: test/test_bc.py [hunk 3]
```diff
@@ -204,3 +200,280 @@ def bnd_norm(vec):
...
+    from mirgecom.boundary import IsothermalNoSlipBoundary
+
+    gas_model = GasModel(eos=IdealSingleGas(gas_const=1.0),
+                         transport=SimpleTransport(viscosity=sigma,
+                                                   thermal_conductivity=kappa))
+
+    wall = IsothermalNoSlipBoundary(wall_temperature=wall_temp)
+
+    # from pytools.convergence import EOCRecorder
+    # eoc = EOCRecorder()
```
Comment (thomasgibson) [inline test/test_bc.py:239]:
```suggestion
```

File: test/test_bc.py [hunk 4]
```diff
@@ -204,3 +200,280 @@ def bnd_norm(vec):
...
+
+    #    for np1 in [4, 8, 12]:
```
Comment (thomasgibson) [inline test/test_bc.py:241]:
```suggestion
```

File: test/test_bc.py [hunk 5]
```diff
@@ -204,3 +200,280 @@ def bnd_norm(vec):
...
+    npts_geom = 17
+    a = 1.0
+    b = 2.0
+    mesh = _get_box_mesh(dim=dim, a=a, b=b, n=npts_geom)
+    #    boundaries = {BTAG_ALL: wall}
+    # for i in range(dim):
+    #     boundaries[DTAG_BOUNDARY("-"+str(i+1))] = 0
+    #     boundaries[DTAG_BOUNDARY("+"+str(i+1))] = 0
```
Comment (thomasgibson) [inline test/test_bc.py:249]:
```suggestion
```

File: test/test_bc.py [hunk 6]
```diff
@@ -204,3 +200,280 @@ def bnd_norm(vec):
...
+
+    discr = EagerDGDiscretization(actx, mesh, order=order)
+    nodes = thaw(actx, discr.nodes())
+    nhat = thaw(actx, discr.normal(BTAG_ALL))
+    print(f"{nhat=}")
+    # h = 1.0 / np1
```
Comment (thomasgibson) [inline test/test_bc.py:255]:
```suggestion
```

File: test/test_bc.py [hunk 7]
```diff
@@ -204,3 +200,280 @@ def bnd_norm(vec):
...
+
+    def scalar_flux_interior(int_tpair):
+        normal = thaw(actx, discr.normal(int_tpair.dd))
+        # Hard-coding central per [Bassi_1997]_ eqn 13
+        flux_weak = gradient_flux_central(int_tpair, normal)
+        return discr.project(int_tpair.dd, "all_faces", flux_weak)
+
+    # utility to compare stuff on the boundary only
+    # from functools import partial
+    # bnd_norm = partial(discr.norm, p=np.inf, dd=BTAG_ALL)
```
Comment (thomasgibson) [inline test/test_bc.py:267]:
```suggestion
```

File: test/test_navierstokes.py [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+    bttf = {}
+    for i in range(dim):
+        bttf["-"+str(i+1)] = ["-"+dim_names[i]]
+        bttf["+"+str(i+1)] = ["+"+dim_names[i]]
+    from meshmode.mesh.generation import generate_regular_rect_mesh as gen
+    return gen(a=a, b=b, n=n, boundary_tag_to_face=bttf, mesh_type=t)
+
+
+@pytest.mark.parametrize("order", [2, 3])
+def test_poiseuille_rhs(actx_factory, order):
```
Comment (thomasgibson) [inline test/test_navierstokes.py:215]: This test is a good candidate to test overintegration. I would try:
```suggestion
@pytest.mark.parametrize("order", [2, 3])
@pytest.mark.parametrize("overintegration", [False, True])
def test_poiseuille_rhs(actx_factory, order, overintegration):
```
And then handle the `overintegration` flag appropriately

File: test/test_navierstokes.py [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        npts_axis = nfac*(12, 20)
+        box_ll = (left_boundary_location, ybottom)
+        box_ur = (right_boundary_location, ytop)
+        mesh = _get_box_mesh(2, a=box_ll, b=box_ur, n=npts_axis)
+
+        logger.info(
+            f"Number of {dim}d elements: {mesh.nelements}"
+        )
+
+        discr = EagerDGDiscretization(actx, mesh, order=order)
```
Comment (thomasgibson) [inline test/test_navierstokes.py:287]: Use this as a reference to update this according to `overintegration`: https://github.com/illinois-ceesd/mirgecom/blob/thg/flux-diff-temp-seeding/examples/pulse-mpi.py#L164-L177

**Review by thomasgibson — COMMENTED** [review]
Thanks for doing this! Here are some initial comments. Also, make sure to update `test_bc.py` to include some verification tests

File: mirgecom/inviscid.py [hunk 10]
```diff
@@ -84,30 +82,102 @@ def inviscid_flux(state):
...
+    rho_int = state_pair.int.mass_density
+    rho_ext = state_pair.ext.mass_density
+    c_int = state_pair.int.speed_of_sound
+    c_ext = state_pair.ext.speed_of_sound
+
+    p_star = (0.5*(p_int + p_ext) + (1./8.)*(u_int - u_ext)
+              * (rho_int + rho_ext) * (c_int + c_ext))
+
+    gamma_int = gas_model.eos.gamma(state_pair.int.cv)
+    gamma_ext = gas_model.eos.gamma(state_pair.ext.cv)
```
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/inviscid.py:143]: Nice catch!!

File: mirgecom/flux.py [hunk 11]
```diff
@@ -35,6 +44,40 @@
...
+    """Lax-Friedrichs/Rusanov low-level numerical flux."""
+    return (f_minus + f_plus - lam*(q_plus - q_minus))/2
+
+
+def num_flux_central(f_minus, f_plus, **kwargs):
+    """Central low-level numerical flux."""
+    return (f_plus + f_minus)/2
+
+
+def hll_flux_driver(state_pair, physical_flux_func,
```
Comment (inducer) [inline mirgecom/flux.py:57]: How come this is not called `num_flux_hll`? How come its interface is inconsistent with the other fluxes?

File: mirgecom/inviscid.py [hunk 12]
```diff
@@ -84,30 +83,102 @@ def inviscid_flux(state):
                           momentum=mom_flux, species_mass=species_mass_flux)
 
 
-def inviscid_facial_flux(discr, state_tpair, local=False):
-    r"""Return the flux across a face given the solution on both sides *q_tpair*.
+def inviscid_flux_rusanov(state_pair, gas_model, normal, **kwargs):
```
Comment (inducer) [inline mirgecom/inviscid.py:86]: Why the naming inconsistency of `_lfr` vs `_rusanov`?

File: mirgecom/flux.py [hunk 13]
```diff
@@ -35,6 +44,40 @@
...
+                    s_minus, s_plus, normal):
+    """State-to-flux driver for hll numerical fluxes."""
```
Comment (inducer) [inline mirgecom/flux.py:59]: This should cite the research paper on which it is based.

File: mirgecom/flux.py [hunk 14]
```diff
@@ -35,6 +44,40 @@
...
+    actx = state_pair.int.array_context
+    zeros = 0.*state_pair.int.mass_density
+
+    f_minus = physical_flux_func(state_pair.int)@normal
+    f_plus = physical_flux_func(state_pair.ext)@normal
+    q_minus = state_pair.int.cv
+    q_plus = state_pair.ext.cv
```
Comment (inducer) [inline mirgecom/flux.py:66]: If this is intended to be generic across conservation laws, it should not access `.cv`.

File: mirgecom/flux.py [hunk 15]
```diff
@@ -35,6 +44,40 @@
...
+    return (f_minus + f_plus - lam*(q_plus - q_minus))/2
+
+
+def num_flux_central(f_minus, f_plus, **kwargs):
+    """Central low-level numerical flux."""
+    return (f_plus + f_minus)/2
+
+
+def hll_flux_driver(state_pair, physical_flux_func,
+                    s_minus, s_plus, normal):
```
Comment (inducer) [inline mirgecom/flux.py:58]: * Document what `{f,s}_{plus,minus}` are.

**Review by inducer — COMMENTED** [review]
Thanks for working on this. Some thoughts from a quick scroll.

File: test/test_navierstokes.py (hunk 8 above)
Comment (MTCam) [reply test/test_navierstokes.py:215]: I'm working on this @thomasgibson.   I'm having troubles with the Poiseuille test, and will ping you as soon as I have it working right.

File: mirgecom/flux.py (hunk 15 above)
Comment (MTCam) [reply mirgecom/flux.py]: Ok, see what you think of its current state.  I rearranged it to be consistent with LFR and more encapsulation of concerns.  The docstring was corrected to match the code, and is mostly self-explanatory - but needs more detail with references.

File: mirgecom/flux.py (hunk 14 above)
Comment (MTCam) [reply mirgecom/flux.py]: Fixed, I think.

File: mirgecom/inviscid.py (hunk 12 above)
Comment (MTCam) [reply mirgecom/inviscid.py:86]: The intent is that the underlying LFR numerical flux can be switched from Rusanov proper to general Lax-Friedrichs by the choice of the penalty factor "lambda".   The Rusanov-specific one chooses the max local wavespeed.

File: mirgecom/flux.py (hunk 11 above)
Comment (MTCam) [reply mirgecom/flux.py]: Fixed I think.

File: mirgecom/flux.py (hunk 13 above)
Comment (MTCam) [reply mirgecom/flux.py]: Fixed up I think.

**Comment by MTCam** [comment]
See how this grabs you @majosm.   ([3f4135f](https://github.com/illinois-ceesd/mirgecom/pull/581/commits/3f4135f4c8409892abbd037700519b8363e92ac5))

**Comment by inducer** [comment]
- Does it make sense to look at this before (MMS) verification is in place?
- There are 702 commits here...  could those be squashed into more sensible units?

**Comment by MTCam** [comment]
> * Does it make sense to look at this before (MMS) verification is in place?

I don't think it makes sense to continue delaying this which we have been using since the beginning.   I suggest we review the code and do any iterations that reviewers require on the code mechanics and if MMS support comes online before we merge, then we'll deal with it.

> * There are 702 commits here...  could those be squashed into more sensible units?

May I ask why the number of commits is relevant for review?  Can we not evaluate the current state of it that is trying to merge?  The longer we let this sit (more than a year now), the more commits will pile up as we keep it up-to-date with all the other things that need to merge for the prediction.

**Comment by inducer** [comment]
> I don't think it makes sense to continue delaying this which we have been using since the beginning.

Believe me, I don't want to delay this. I'm happy to review for structure, but honestly I don't think it's a good use of my time to review for correctness before sufficient verification is in place. Both parts of review are needded

> May I ask why the number of commits is relevant for review?

- It slows down the Github interface. (And this is already extremely unpleasant to work with. Try adding a review comment. On my machine, each key press in a new review comment incurs a one-second delay.)
- It makes the discussion unwieldy (in that previous comments disappear under layers of "show previous")
- We can't possibly intend to hit "merge" on this in its current state ("squash" at best), so it doesn't represent the state that will wind up committed.

File: mirgecom/navierstokes.py [hunk 16]
```diff
@@ -0,0 +1,416 @@
...
+
+    if grad_cv is None:
+        grad_cv = grad_cv_operator(
+            discr, gas_model, boundaries, state, time=time,
+            numerical_flux_func=gradient_numerical_flux_func,
+            quadrature_tag=quadrature_tag,
+            operator_states_quad=operator_states_quad)
+
+    # Communicate grad(CV) and put it on the quadrature domain
+    # FIXME/ReviewQuestion: communicate grad_cv - already on quadrature dom?
```
Comment (inducer) [inline mirgecom/navierstokes.py:350]: What's your question here? If it is whether values on the quadrature discretization can be commuicated: I believe that's possible, but I'd prefer if we don't, because that implicitly rules out non-conforming interfaces.

File: mirgecom/navierstokes.py [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+            interior=op.project(discr, local_dd, local_dd_quad, utpair.int),
+            exterior=op.project(discr, local_dd, local_dd_quad, utpair.ext)
+        )
+
+    # }}}
+
+    # {{{ === Compute grad(CV) ===
+
+    if grad_cv is None:
+        grad_cv = grad_cv_operator(
```
Comment (inducer) [inline mirgecom/navierstokes.py:343]: Why are there separate gradient functions for CV and temparature? A gradient is a gradient, right?

File: mirgecom/navierstokes.py [hunk 18]
```diff
@@ -0,0 +1,416 @@
...
+
+        # All surface contributions from the inviscid fluxes
+        - inviscid_boundary_flux_for_divergence_operator(
+            discr, gas_model, boundaries, inter_elem_bnd_states_quad,
+            domain_bnd_states_quad, quadrature_tag=quadrature_tag,
+            numerical_flux_func=inviscid_numerical_flux_func, time=time)
+
+    )
+
+    return div_operator(discr, dd_vol_quad, dd_faces_quad, vol_term, bnd_term)
```
Comment (inducer) [inline mirgecom/navierstokes.py:414]: I still find that the use of `div_operator` here obscures what's really going on. But that's not for this review, I guess.

File: mirgecom/navierstokes.py [hunk 19]
```diff
@@ -0,0 +1,416 @@
...
+    -------
+    :class:`numpy.ndarray`
+
+        Array of :class:`~meshmode.dof_array.DOFArray` representing the gradient of
+        the fluid temperature.
+    """
+    dd_vol_quad = DOFDesc("vol", quadrature_tag)
+    dd_faces_quad = DOFDesc("all_faces", quadrature_tag)
+
+    if operator_states_quad is None:
```
Comment (inducer) [inline mirgecom/navierstokes.py:220]: Is there a situation where this *wouldn't* be passed?

File: mirgecom/boundary.py [hunk 20]
```diff
@@ -38,91 +40,264 @@
...
-        self._inviscid_bnd_flux_func = inviscid_boundary_flux_func
-        self._inviscid_div_flux_func = inviscid_facial_flux_func
+        self._inviscid_flux_func = inviscid_flux_func
+        self._temperature_grad_flux_func = temperature_gradient_flux_func
         self._bnd_temperature_func = boundary_temperature_func
+        self._grad_num_flux_func = gradient_numerical_flux_func
+        self._cv_gradient_flux_func = cv_gradient_flux_func
+        self._viscous_flux_func = viscous_flux_func
+        self._bnd_grad_cv_func = boundary_gradient_cv_func
+        self._bnd_grad_temperature_func = boundary_gradient_temperature_func
```
Comment (inducer) [inline mirgecom/boundary.py:128]: A struct of callback functions is not lovely: Use an abstract base class. There's even (sort of, hand-coded) inheritance logic below!

File: mirgecom/boundary.py [hunk 21]
```diff
@@ -38,91 +40,264 @@
...
+                            interior=state_minus.cv,
+                            exterior=boundary_state.cv)
+
+        actx = state_minus.array_context
+        nhat = thaw(discr.normal(btag), actx)
+        return self._boundary_quantity(
+            discr, btag=btag,
+            quantity=self._grad_num_flux_func(cv_pair, nhat), **kwargs)
+
+    def _gradient_flux_for_prescribed_temperature(self, discr, btag, gas_model,
```
Comment (inducer) [inline mirgecom/boundary.py:210]: Comment on what "for prescribed temperature" means here. (Docstring/comment)

File: mirgecom/boundary.py [hunk 22]
```diff
@@ -38,91 +40,264 @@
...
+    def _identical_state(self, state_minus, **kwargs):
+        return state_minus
+
+    def _identical_grad_cv(self, grad_cv_minus, **kwargs):
+        return grad_cv_minus
+
+    def _identical_grad_temperature(self, grad_t_minus, **kwargs):
+        return grad_t_minus
+
+    def _gradient_flux_for_prescribed_cv(self, discr, btag, gas_model, state_minus,
```
Comment (inducer) [inline mirgecom/boundary.py:193]: What type is `state_minus`?

File: mirgecom/boundary.py [hunk 23]
```diff
@@ -38,91 +40,264 @@
...
+                self._gradient_flux_for_prescribed_temperature
+
+        if not self._viscous_flux_func:
+            self._viscous_flux_func = self._viscous_flux_for_prescribed_state
+        if not self._bnd_grad_cv_func:
+            self._bnd_grad_cv_func = self._identical_grad_cv
+        if not self._bnd_grad_temperature_func:
+            self._bnd_grad_temperature_func = self._identical_grad_temperature
 
     def _boundary_quantity(self, discr, btag, quantity, **kwargs):
```
Comment (inducer) [inline mirgecom/boundary.py:159]: This takes in arbitary kwargs without checking their validity, inviting typos.

**Review by inducer — COMMENTED** [review]
Some initial thoughts here, all I have time for for the moment. Also: structural review only, did not look at correctness.

File: .github/workflows/ci.yaml [hunk 24]
```diff
@@ -77,8 +77,8 @@ jobs:
             MINIFORGE_INSTALL_DIR=.miniforge3
             . "$MINIFORGE_INSTALL_DIR/bin/activate" testing
             cd test
-            python -m pip install pytest-xdist
-            python -m pytest -n auto --cov=mirgecom --durations=0 --tb=native --junitxml=pytest.xml --doctest-modules -rxsw . ../doc/*.rst ../doc/*/*.rst
```
[the remark this answers is not in the document]
Comment (MTCam) [reply .github/workflows/ci.yaml:81]: xdist runs CI out of memory.

File: mirgecom/flux.py [hunk 25]
```diff
@@ -35,179 +40,203 @@
...
+        object array of :class:`~meshmode.dof_array.DOFArray` with the
+        central numerical flux.
     """
-    return trace_pair.avg@normal
+    return (f_plus + f_minus)/2
 
 
-def divergence_flux_lfr(cv_tpair, f_tpair, normal, lam):
-    r"""Compute Lax-Friedrichs/Rusanov flux after [Hesthaven_2008]_, Section 6.6.
+def num_flux_hll(f_minus, f_plus, q_minus, q_plus, s_minus, s_plus):
```
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/flux.py:111]: 👍  I will move it.

File: mirgecom/navierstokes.py (hunk 16 above)
Comment (MTCam) [reply mirgecom/navierstokes.py:350]: This question was from @majosm to me.   I think we came to the same conclusion; we _could_ but probably shouldn't.

File: mirgecom/navierstokes.py (hunk 19 above)
Comment (MTCam) [reply mirgecom/navierstokes.py:220]: Yes.  In almost all situations, NS computes its own operator states.  The exception is those cases where the operator states have already been computed by another component (like in wall model runs).  This was part of a refactor to support the wall model.

File: mirgecom/boundary.py (hunk 20 above)
Comment (MTCam) [reply mirgecom/boundary.py:128]: We are aware of this issue and plan to refactor the boundary treatments accordingly - based on previous feedback from @inducer and @majosm.   Is that something we want to tackle in this PR?

File: .github/workflows/ci.yaml (hunk 24 above)
[the remark this answers is not in the document]
Comment (inducer) [reply .github/workflows/ci.yaml:81]: #647

File: mirgecom/navierstokes.py [hunk 26]
```diff
@@ -0,0 +1,416 @@
...
+
+
+def ns_operator(discr, gas_model, state, boundaries, *, time=0.0,
+                inviscid_numerical_flux_func=inviscid_flux_rusanov,
+                gradient_numerical_flux_func=gradient_flux_central,
+                viscous_numerical_flux_func=viscous_flux_central,
+                quadrature_tag=DISCR_TAG_BASE,
+                # Added to avoid repeated computation
+                # FIXME: See if there's a better way to do this
+                operator_states_quad=None,
```
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/navierstokes.py:269]: When other components compute the operator states prior to calling NS.

File: mirgecom/boundary.py (hunk 20 above)
Comment (inducer) [reply mirgecom/boundary.py:128]: Bump to issue is OK with me.

File: mirgecom/navierstokes.py (hunk 17 above)
Comment (majosm) [reply mirgecom/navierstokes.py:343]: The two functions interact with the fluid state and boundary conditions differently. It could see it being possible to create a single `gradient_operator` that could be used for both, but that's probably a story for another PR...

File: mirgecom/boundary.py (hunk 20 above)
Comment (MTCam) [reply mirgecom/boundary.py:128]: #576 is the issue created for this around the time we last discussed it.

File: mirgecom/navierstokes.py (hunk 19 above)
Comment (majosm) [reply mirgecom/navierstokes.py:220]: And for gradients: visualization

File: mirgecom/flux.py [hunk 27]
```diff
@@ -35,179 +40,203 @@
...
 
     Parameters
     ----------
-    cv_tpair: :class:`~grudge.trace_pair.TracePair`
+    f_minus
+        Normal component of physical flux interior to (left of) interface
 
-        Solution trace pair for faces for which numerical flux is to be calculated
+    f_plus
+        Normal component of physical flux exterior to (right of) interface
```
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/flux.py]: ([81cca0f](https://github.com/illinois-ceesd/mirgecom/pull/581/commits/81cca0f74abb57806925cbaa32326e9faa9229fa))

File: mirgecom/navierstokes.py (hunk 18 above)
Comment (MTCam) [reply mirgecom/navierstokes.py:414]: (https://github.com/illinois-ceesd/mirgecom/issues/598)

File: mirgecom/boundary.py (hunk 21 above)
Comment (MTCam) [reply mirgecom/boundary.py:210]: This function will be eliminated on BC refactor.   I will comment it up to try to make it more clear in the interim.

File: mirgecom/navierstokes.py [hunk 28]
```diff
@@ -0,0 +1,416 @@
...
+        operator_states_quad = make_operator_fluid_states(
+            discr, state, gas_model, boundaries, quadrature_tag)
+
+    vol_state_quad, inter_elem_bnd_states_quad, domain_bnd_states_quad = \
+        operator_states_quad
+
+    # {{{ Local utilities
+
+    # transfer trace pairs to quad grid, update pair dd
+    def _interp_to_surf_quad(utpair):
```
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/navierstokes.py:329]: Can we chat about how to use inducer/grudge#254 tomorrow?

File: mirgecom/boundary.py (hunk 23 above)
Comment (MTCam) [reply mirgecom/boundary.py]: ([b0a4356](https://github.com/illinois-ceesd/mirgecom/pull/581/commits/b0a4356683cb04e161ef951ceb9d6094241d6992)

File: mirgecom/boundary.py (hunk 22 above)
Comment (MTCam) [reply mirgecom/boundary.py:331]: ([8344799](https://github.com/illinois-ceesd/mirgecom/pull/581/commits/8344799ab04e877403b0e5162d2f6f9ece3505b4))

File: mirgecom/navierstokes.py (hunk 17 above)
Comment (MTCam) [reply mirgecom/navierstokes.py:343]: These are the CV-specific and fluid-temperature-specific handlers to compute the gradients of those quantities.  Underneath, they both use the gradient-is-a-gradient interface `gradient_operator`.

File: mirgecom/boundary.py (hunk 21 above)
Comment (MTCam) [reply mirgecom/boundary.py:350]: ([8344799](https://github.com/illinois-ceesd/mirgecom/pull/581/commits/8344799ab04e877403b0e5162d2f6f9ece3505b4))

File: mirgecom/flux.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/flux.py]: #648

File: test/test_navierstokes.py [hunk 29]
```diff
@@ -0,0 +1,375 @@
...
+    return gen(a=a, b=b, n=n, boundary_tag_to_face=bttf, mesh_type=t)
+
+
+@pytest.mark.parametrize("order", [2, 3])
+def test_poiseuille_rhs(actx_factory, order):
+    """Test the Navier-Stokes operator using a Poiseuille state.
+
+    This state should yield rhs = 0 to FP.  The test is performed for 1, 2,
+    and 3 dimensions, with orders 1, 2, and 3, with and without passive species.
+    """
```
Comment (inducer) [inline test/test_navierstokes.py:232]: Some of the details/analysis from https://notes.tiker.net/IQb-yrlFT82tKscEFkhrxA should probably make it in here.

File: test/test_navierstokes.py [hunk 30]
```diff
@@ -0,0 +1,375 @@
...
+        eoc_rec.add_data_point(1.0 / nfac, err_max)
+
+    logger.info(
+        f"V != 0 Errors:\n{eoc_rec}"
+    )
+
+    assert (
+        eoc_rec.order_estimate() >= order - 0.5
+        or eoc_rec.max_error() < tol_fudge
+    )
```
Comment (inducer) [inline test/test_navierstokes.py:375]:
```suggestion
    if order <= 1:
        assert eoc_rec.order_estimate() >= order - 0.5
    else:
        # Poiseuille is a quadratic profile, exactly represented by quadratic and higher
        assert eoc_rec.max_error() < tol_fudge
```

File: mirgecom/flux.py (hunk 11 above)
Comment (MTCam) [reply mirgecom/flux.py]: Resolving this as we've addressed in #648

File: mirgecom/flux.py (hunk 13 above)
Comment (MTCam) [reply mirgecom/flux.py]: #648

File: mirgecom/flux.py (hunk 14 above)
Comment (MTCam) [reply mirgecom/flux.py]: #648

File: mirgecom/navierstokes.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (MTCam) [reply mirgecom/navierstokes.py]: Fixed up!
